pkg/apis/autoscaling/v1alpha1/cronfederatedhpa_types.go

```diff
@@ -0,0 +1,218 @@
+	// +required
+	Rules []CronFederatedHPARule `json:"rules"`
+}
+
+// CronFederatedHPARule declares a schedule as well as scale actions.
+type CronFederatedHPARule struct {
+	// Name of the rule.
+	// Each rule in a CronFederatedHPA must have a unique name.
+	//
+	// Note: the name will be used as an identifier to record its execution
+	// history. Changing the name will be considered as deleting the old rule
+	// and adding a new rule, that means the original execution history will be
+	// discarded.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=32
+	// +required
+	Name string `json:"name"`
+
+	// Schedule is the cron expression that represents a periodical time.
+	// The syntax follows https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#schedule-syntax.
+	// +required
+	Schedule string `json:"schedule"`
+
+	// TargetReplicas is the target replicas to be scaled for resources
+	// referencing by ScaleTargetRef of this CronFederatedHPA.
+	// Only needed when referencing resource is not FederatedHPA.
+	// +optional
+	TargetReplicas *int32 `json:"targetReplicas,omitempty"`
+
+	// TargetMinReplicas is the target MinReplicas to be set for FederatedHPA.
+	// Only needed when referencing resource is FederatedHPA.
+	// TargetMinReplicas and TargetMaxReplicas can be specified together or
+	// either one can be specified alone.
+	// nil means the MinReplicas(.spec.minReplicas) of the referencing FederatedHPA
+	// will not be updated.
+	// +optional
+	TargetMinReplicas *int32 `json:"targetMinReplicas,omitempty"`
+
+	// TargetMaxReplicas is the target MaxReplicas to be set for FederatedHPA.
+	// Only needed when referencing resource is FederatedHPA.
+	// TargetMinReplicas and TargetMaxReplicas can be specified together or
+	// either one can be specified alone.
+	// nil means the MaxReplicas(.spec.maxReplicas) of the referencing FederatedHPA
+	// will not be updated.
+	// +optional
+	TargetMaxReplicas *int32 `json:"targetMaxReplicas,omitempty"`
+
+	// Suspend tells the controller to suspend subsequent executions.
+	// Defaults to false.
+	//
+	// +kubebuilder:default=false
+	// +optional
+	Suspend *bool `json:"suspend,omitempty"`
+
+	// TimeZone for the giving schedule.
+	// If not specified, this will default to the time zone of the
+	// karmada-controller-manager process.
+	// Invalid TimeZone will be rejected when applying by karmada-webhook.
+	// see https://en.wikipedia.org/wiki/List_of_tz_database_time_zones for the
+	// all timezones.
+	// +optional
+	TimeZone *string `json:"timeZone,omitempty"`
+
+	// SuccessfulHistoryLimit represents the count of successful execution items
+	// for each rule.
+	// The value must be a positive integer. It defaults to 3.
+	//
+	// +kubebuilder:default=3
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=32
+	// +optional
+	SuccessfulHistoryLimit *int32 `json:"successfulHistoryLimit,omitempty"`
```

Comment:
There is no situation where it is `nil`.

I think that `int32` may be better than `*int32`.

Reply:
In fact, we don't need pointer type for `Suspend`, `TimeZone`, `SuccessfulHistoryLimit`, and `FailedHistoryLimit`, because they all have default values. 

For `FailedHistoryLimit`, we don't care whether it is provided. 

@jwcesign What's your preference?

Reply:
In my opinion, it would be better to have the optional field as a pointer since this is a commonly accepted convention. Therefore, I prefer to maintain it as a pointer type.

Reply:
Fine. We can review this issue after the implementation, and even if we are going to change the type, it still is a compatible change.

Reply:
@chaunceyjiang how about that?

Reply:
ok.